### PR TITLE
WebSocket API improvements

### PR DESCRIPTION
**Pull request by ViacheslavKlimov**
## Pull Request description

Put your PR description here instead of this sentence.   

## General checklist

- [ ] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [ ] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [ ] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [ ] Description contains human-readable scope of changes.
- [ ] Description contains brief notes about what needs to be added to the documentation.
- [ ] No merge conflicts, commented blocks of code, code formatting issues.
- [ ] Changes are backward compatible or upgrade script is provided.
- [ ] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [ ] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [ ] If new dependency was added: the dependency tree is checked for conflicts.
- [ ] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [ ] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [ ] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/service/ws/DefaultWebSocketService.java [hunk 1]
```diff
@@ -234,25 +231,27 @@ public void handleWebSocketMsg(WebSocketSessionRef sessionRef, String msg) {
...
+            return;
+        }
+        String sessionId = sessionRef.getSessionId();
+        for (WsCmdsHandler<? extends WsCmd> cmdHandler : cmdsHandlers) {
+            List<? extends WsCmd> cmds = cmdHandler.extract(cmdsWrapper);
+            if (cmds == null) {
+                continue;
+            }
+            for (WsCmd cmd : cmds) {
                 if (validateSessionMetadata(sessionRef, cmd.getCmdId(), sessionId)) {
```
Comment (ViacheslavKlimov) [inline application/src/main/java/org/thingsboard/server/service/ws/DefaultWebSocketService.java:255]: No need to validate session metadata for each cmd, as it is the same session

File: application/src/main/java/org/thingsboard/server/controller/plugin/TbWebSocketHandler.java [hunk 2]
```diff
@@ -97,26 +112,89 @@ public class TbWebSocketHandler extends TextWebSocketHandler implements WebSocke
...
-                webSocketService.handleWebSocketMsg(sessionMd.sessionRef, message.getPayload());
-            } else {
+            SessionMetaData sessionMd = getSessionMd(session.getId());
+            if (sessionMd == null) {
                 log.trace("[{}] Failed to find session", session.getId());
                 session.close(CloseStatus.SERVER_ERROR.withReason("Session not found!"));
+                return;
             }
+            String msg = message.getPayload();
+            sessionMd.onMsg(msg);
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/controller/plugin/TbWebSocketHandler.java:138]: sessionMd.onMsg(message.getPayload())

File: application/src/main/java/org/thingsboard/server/controller/plugin/TbWebSocketHandler.java [hunk 3]
```diff
@@ -97,26 +112,89 @@ public class TbWebSocketHandler extends TextWebSocketHandler implements WebSocke
     private final ConcurrentMap<UserId, Set<String>> regularUserSessionsMap = new ConcurrentHashMap<>();
     private final ConcurrentMap<UserId, Set<String>> publicUserSessionsMap = new ConcurrentHashMap<>();
 
+    private final Cache<String, SessionMetaData> pendingSessions = Caffeine.newBuilder()
+            .expireAfterWrite(10, TimeUnit.SECONDS)
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/controller/plugin/TbWebSocketHandler.java:116]: move to yml please

File: application/src/main/java/org/thingsboard/server/controller/plugin/TbWebSocketHandler.java [hunk 4]
```diff
@@ -97,26 +112,89 @@ public class TbWebSocketHandler extends TextWebSocketHandler implements WebSocke
...
+                    cmdsWrapper = JacksonUtil.fromString(msg, TelemetryCmdsWrapper.class).toCommonCmdsWrapper();
+                    break;
+                case NOTIFICATIONS:
+                    cmdsWrapper = JacksonUtil.fromString(msg, NotificationCmdsWrapper.class).toCommonCmdsWrapper();
+                    break;
+                default:
+                    return;
+            }
+        } catch (Exception e) {
+            log.warn("Failed to decode subscription cmd: {}", e.getMessage(), e);
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/controller/plugin/TbWebSocketHandler.java:162]: add tenant id + user id + session id to be able to detect the user (if any)

File: application/src/main/java/org/thingsboard/server/controller/plugin/TbWebSocketHandler.java [hunk 5]
```diff
@@ -444,10 +575,10 @@ private boolean checkLimits(WebSocketSession session, WebSocketSessionRef sessio
                     }
                 }
                 if (!limitAllowed) {
-                        log.info("[{}][{}][{}] Failed to start session. Max regular user sessions limit reached"
-                                , sessionRef.getSecurityCtx().getTenantId(), sessionRef.getSecurityCtx().getId(), sessionId);
-                        session.close(CloseStatus.POLICY_VIOLATION.withReason("Max regular user sessions limit reached"));
-                        return false;
+                    log.info("[{}][{}][{}] Failed to start session. Max regular user sessions limit reached"
+                            , sessionRef.getSecurityCtx().getTenantId(), sessionRef.getSecurityCtx().getId(), sessionId);
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/controller/plugin/TbWebSocketHandler.java:579]: refactor into toString method for sessionRef to be null safe in any case.

File: application/src/main/java/org/thingsboard/server/service/ws/WsCommandsWrapper.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            @Type(name = "NOTIFICATIONS_COUNT", value = NotificationsCountSubCmd.class),
+            @Type(name = "MARK_NOTIFICATIONS_AS_READ", value = MarkNotificationsAsReadCmd.class),
+            @Type(name = "MARK_ALL_NOTIFICATIONS_AS_READ", value = MarkAllNotificationsAsReadCmd.class),
+            @Type(name = "ALARM_DATA_UNSUBSCRIBE", value = AlarmDataUnsubscribeCmd.class),
+            @Type(name = "ALARM_COUNT_UNSUBSCRIBE", value = AlarmCountUnsubscribeCmd.class),
+            @Type(name = "ENTITY_DATA_UNSUBSCRIBE", value = EntityDataUnsubscribeCmd.class),
+            @Type(name = "ENTITY_COUNT_UNSUBSCRIBE", value = EntityCountUnsubscribeCmd.class),
+            @Type(name = "NOTIFICATIONS_UNSUBSCRIBE", value = NotificationsUnsubCmd.class),
+    })
+    private List<WsCmd> cmds;
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/service/ws/WsCommandsWrapper.java:69]: Please make sure that we can process a message that contains both authCmd and multiple WsCmd.

**Pull request merged**